File: services/api/src/webwoven_api/domain/hints.py
"""Deterministic, graph-grounded hint selection."""

from collections.abc import Iterable
from dataclasses import dataclass
from enum import StrEnum

from webwoven_api.domain.errors import DomainError


class HintType(StrEnum):
    COMPASS = "compass"
    LENS = "lens"
    MAP_FRAGMENT = "map_fragment"


class HintOutcome(StrEnum):
    PROMISING = "promising"
    LONGER = "longer"
    DEAD_END = "dead_end"


HINT_PENALTIES: dict[HintType, int] = {
    HintType.COMPASS: 75,
    HintType.LENS: 150,
    HintType.MAP_FRAGMENT: 250,
}


@dataclass(frozen=True, slots=True)
class HintCandidate:
    relation_key: str
    entity_id: str
    entity_label: str
    distance: int | None


@dataclass(frozen=True, slots=True)
class HintResult:
    hint_type: HintType
    penalty: int
    relation_key: str | None
    entity_id: str | None
    message: str
    outcome: HintOutcome | None = None


def select_hint(
    hint_type: HintType,
    candidates: Iterable[HintCandidate],
    *,
    selected_relation_key: str | None = None,
    selected_entity_id: str | None = None,
) -> HintResult:
    """Choose a hint using only precomputed graph distances and stable sorting."""
    available = tuple(candidates)
    if hint_type is HintType.COMPASS:
        if selected_relation_key is None or selected_entity_id is None:
            raise DomainError("route_required", "Choose a specific route for the Compass.")
        selected = next(
            (
                candidate
                for candidate in available
                if candidate.relation_key == selected_relation_key
                and candidate.entity_id == selected_entity_id
            ),
            None,
        )
        if selected is None:
            raise DomainError("hint_unavailable", "That route is no longer available.")
        reachable = tuple(candidate for candidate in available if candidate.distance is not None)
        best = min(reachable, key=_candidate_sort_key) if reachable else None
        if selected.distance is None:
            outcome = HintOutcome.DEAD_END
            message = f"Compass: {selected.entity_label} is a dead end from here."
        elif best is not None and selected.distance == best.distance:
            outcome = HintOutcome.PROMISING
            message = f"Compass: {selected.entity_label} is a promising route."
        else:
            outcome = HintOutcome.LONGER
            message = f"Compass: {selected.entity_label} takes a longer route."
        return HintResult(
            hint_type=hint_type,
            penalty=HINT_PENALTIES[hint_type],
            relation_key=selected.relation_key,
            entity_id=selected.entity_id,
            message=message,
            outcome=outcome,
        )

    reachable = tuple(candidate for candidate in available if candidate.distance is not None)
    if not reachable:
        raise DomainError("hint_unavailable", "No grounded hint is available from this entity.")
    best = min(reachable, key=_candidate_sort_key)

    if hint_type is HintType.LENS:
        return HintResult(
            hint_type=hint_type,
            penalty=HINT_PENALTIES[hint_type],
            relation_key=best.relation_key,
            entity_id=best.entity_id,
            message=f"Lens: {best.entity_label} is on a near-optimal route.",
            outcome=HintOutcome.PROMISING,
        )

    return HintResult(
        hint_type=hint_type,
        penalty=HINT_PENALTIES[hint_type],
        relation_key=best.relation_key,
        entity_id=best.entity_id,
        message=f"Map Fragment: {best.entity_label} is a valid bridge ahead.",
        outcome=HintOutcome.PROMISING,
    )


def _candidate_sort_key(candidate: HintCandidate) -> tuple[int, str, str]:
    assert candidate.distance is not None
    return (candidate.distance, candidate.relation_key, candidate.entity_id)

File: apps/web/src/lib/domain/hints.ts
import type {
  HintMarker,
  HintOutcome,
  HintType,
  RelationGroup,
  UsedHint,
} from "../api/types";
import { hintPenalty } from "./scoring";

interface HintContext {
  selectedPropertyId?: string;
  selectedEntityQid?: string;
  distanceToTarget: (entityQid: string) => number | null;
}

interface Candidate {
  group: RelationGroup;
  edge: RelationGroup["edges"][number];
  distance: number | null;
}

export function applyHintToGroups(
  groups: RelationGroup[],
  type: HintType,
  context: HintContext,
): { groups: RelationGroup[]; hint: UsedHint } {
  const candidates = groups.flatMap((group) =>
    group.edges.map((edge) => ({
      group,
      edge,
      distance: context.distanceToTarget(edge.target.qid),
    })),
  );
  const reachable = candidates.filter(
    (candidate): candidate is Candidate & { distance: number } =>
      candidate.distance !== null,
  );
  const best = [...reachable].sort(compareCandidates)[0];

  if (type === "compass") {
    const selected = candidates.find(
      ({ group, edge }) =>
        group.property_id === context.selectedPropertyId &&
        edge.target.qid === context.selectedEntityQid,
    );
    if (!selected) throw new Error("Choose a specific route for the Compass.");
    const outcome: HintOutcome =
      selected.distance === null
        ? "dead_end"
        : selected.distance === best?.distance
          ? "promising"
          : "longer";
    return exactResult(
      groups,
      type,
      selected,
      outcome,
      compassMessage(selected.edge.target.label, outcome),
    );
  }

  if (!best) {
    return {
      groups,
      hint: {
        type,
        penalty: hintPenalty(type),
        message: "No grounded hint is available from this entity.",
      },
    };
  }

  const message =
    type === "lens"
      ? `Lens: ${best.edge.target.label} is on a near-optimal route.`
      : `Map Fragment: ${best.edge.target.label} is a valid bridge ahead.`;
  return exactResult(groups, type, best, "promising", message);
}

function exactResult(
  groups: RelationGroup[],
  type: HintType,
  candidate: Candidate,
  outcome: HintOutcome,
  message: string,
): { groups: RelationGroup[]; hint: UsedHint } {
  return {
    groups: markExactEdge(
      groups,
      candidate.group.property_id,
      candidate.edge.target.qid,
      outcome,
    ),
    hint: {
      type,
      penalty: hintPenalty(type),
      message,
      relation_property_id: candidate.group.property_id,
      entity_qid: candidate.edge.target.qid,
      outcome,
    },
  };
}

function markExactEdge(
  groups: RelationGroup[],
  propertyId: string,
  entityQid: string,
  hint: HintMarker,
): RelationGroup[] {
  return groups.map((group) => ({
    ...group,
    edges: group.edges.map((edge) =>
      group.property_id === propertyId && edge.target.qid === entityQid
        ? { ...edge, hint }
        : edge,
    ),
  }));
}

function compareCandidates(
  left: Candidate & { distance: number },
  right: Candidate & { distance: number },
): number {
  return (
    left.distance - right.distance ||
    left.group.property_id.localeCompare(right.group.property_id) ||
    left.edge.target.qid.localeCompare(right.edge.target.qid)
  );
}

function compassMessage(label: string, outcome: HintOutcome): string {
  if (outcome === "dead_end")
    return `Compass: ${label} is a dead end from here.`;
  if (outcome === "longer") return `Compass: ${label} takes a longer route.`;
  return `Compass: ${label} is a promising route.`;
}
